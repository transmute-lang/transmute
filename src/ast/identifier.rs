use crate::lexer::Span;
use std::fmt::{Display, Formatter};

#[derive(Debug, PartialEq)]
pub struct Identifier {
    name: String,
    span: Span,
}

impl Identifier {
    pub fn new(name: String, span: Span) -> Self {
        Self { name, span }
    }

    pub fn span(&self) -> &Span {
        &self.span
    }

    pub fn name(&self) -> &str {
        &self.name
    }
}

impl Display for Identifier {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.name)
    }
}
