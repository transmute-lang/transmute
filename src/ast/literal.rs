use crate::ast::ids::IdentRefId;
use crate::lexer::Span;
use std::fmt::{Debug, Formatter};

#[derive(Debug, Clone, PartialEq)]
pub struct Literal {
    kind: LiteralKind,
    span: Span,
}

impl Literal {
    pub fn new(kind: LiteralKind, span: Span) -> Self {
        Self { kind, span }
    }

    pub fn kind(&self) -> &LiteralKind {
        &self.kind
    }

    pub fn span(&self) -> &Span {
        &self.span
    }
}

#[derive(Clone, PartialEq)]
pub enum LiteralKind {
    Boolean(bool),
    Identifier(IdentRefId),
    Number(i64),
}

impl Debug for LiteralKind {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            LiteralKind::Boolean(v) => write!(f, "Boolean({v})"),
            LiteralKind::Identifier(v) => write!(f, "IdentRefId({v})"),
            LiteralKind::Number(v) => write!(f, "Number({v})"),
        }
    }
}
