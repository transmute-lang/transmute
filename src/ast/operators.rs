use crate::lexer::{Location, Span};

#[derive(Debug, PartialEq)]
pub struct BinaryOperator {
    kind: BinaryOperatorKind,
    location: Location,
    span: Span,
}

impl BinaryOperator {
    pub fn new(kind: BinaryOperatorKind, location: Location, span: Span) -> Self {
        Self {
            kind,
            location,
            span,
        }
    }

    pub fn kind(&self) -> &BinaryOperatorKind {
        &self.kind
    }
}

#[derive(Debug, PartialEq)]
pub enum BinaryOperatorKind {
    Addition,
}
