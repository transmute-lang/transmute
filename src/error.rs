use crate::lexer::Span;
use std::fmt::{Display, Formatter};

#[derive(Debug)]
pub enum Error {
    // NoMatch,
    UnexpectedChar(Span),
    ExpectedSemicolon(Span),
}

#[derive(Debug, Default, PartialEq)]
pub struct Diagnostics {
    diagnostics: Vec<Diagnostic>,
}

impl Diagnostics {
    pub fn len(&self) -> usize {
        self.diagnostics.len()
    }

    pub fn is_empty(&self) -> bool {
        self.diagnostics.is_empty()
    }

    pub fn iter(&self) -> DiagnosticsIterator {
        DiagnosticsIterator {
            diagnostics: self,
            index: 0,
        }
    }

    pub fn append(&mut self, mut other: Self) {
        self.diagnostics.append(&mut other.diagnostics)
    }

    pub fn push(&mut self, diagnostic: Diagnostic) {
        self.diagnostics.push(diagnostic);
    }

    pub fn report_err<S: Into<String>>(
        &mut self,
        message: S,
        span: Span,
        generated_at: (&'static str, u32),
    ) {
        self.push(Diagnostic::new(message, span, Level::Error, generated_at));
    }
}

impl Display for Diagnostics {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let mut diagnostics = self.diagnostics.to_vec();
        diagnostics.sort_by(|a, b| a.span.cmp(&b.span));

        for diagnostic in diagnostics {
            writeln!(f, " - {}", diagnostic)?;
        }
        Ok(())
    }
}

pub struct DiagnosticsIterator<'a> {
    diagnostics: &'a Diagnostics,
    index: usize,
}

impl<'a> Iterator for DiagnosticsIterator<'a> {
    type Item = &'a Diagnostic;

    fn next(&mut self) -> Option<Self::Item> {
        if self.index < self.diagnostics.len() {
            self.index += 1;
            Some(&self.diagnostics.diagnostics[self.index - 1])
        } else {
            None
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Diagnostic {
    message: String,
    span: Span,
    level: Level,
    generated_at: (&'static str, u32),
}

impl Display for Diagnostic {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        #[cfg(debug_assertions)]
        {
            write!(
                f,
                "[{}:{}] {} at {}:{}",
                self.generated_at.0,
                self.generated_at.1,
                self.message,
                self.span.line(),
                self.span.column()
            )
        }
        #[cfg(not(debug_assertions))]
        {
            write!(
                f,
                "{} at {}:{}",
                self.message,
                self.span.line(),
                self.span.column()
            )
        }
    }
}

impl Diagnostic {
    pub fn new<S: Into<String>>(
        message: S,
        span: Span,
        level: Level,
        generated_at: (&'static str, u32),
    ) -> Self {
        Self {
            message: message.into(),
            span,
            level,
            generated_at,
        }
    }

    pub fn error<S: Into<String>>(
        message: S,
        span: Span,
        generated_at: (&'static str, u32),
    ) -> Self {
        Self::new(message, span, Level::Error, generated_at)
    }

    pub fn message(&self) -> &str {
        &self.message
    }

    pub fn span(&self) -> &Span {
        &self.span
    }

    pub fn level(&self) -> Level {
        self.level
    }
}

#[derive(Debug, Clone, Copy, Ord, PartialOrd, Eq, PartialEq)]
pub enum Level {
    Error,
}
