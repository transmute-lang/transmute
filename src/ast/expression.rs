use crate::ast::literal::Literal;
use crate::ast::operators::{BinaryOperator, UnaryOperator};
use crate::ast::Visitor;
use crate::lexer::{Location, Span};

#[derive(Debug, PartialEq)]
pub struct Expression {
    kind: ExpressionKind,
}

impl Expression {
    pub fn new(kind: ExpressionKind) -> Self {
        Self { kind }
    }

    pub fn kind(&self) -> &ExpressionKind {
        &self.kind
    }

    pub fn location(&self) -> &Location {
        match &self.kind {
            ExpressionKind::Literal(l) => l.location(),
            ExpressionKind::Binary(e, _, _) => e.location(),
            ExpressionKind::Unary(o, _) => o.location(),
        }
    }

    pub fn span(&self) -> Span {
        match &self.kind {
            ExpressionKind::Literal(l) => l.span().clone(),
            ExpressionKind::Binary(l, _, r) => l.span().extend_to(&r.span()),
            ExpressionKind::Unary(o, e) => o.span().extend_to(&e.span()),
        }
    }

    pub fn accept<V, R>(&self, visitor: &mut V) -> R
    where
        V: Visitor<R>,
    {
        visitor.visit_expression(self)
    }
}

impl From<Literal> for Expression {
    fn from(literal: Literal) -> Self {
        Self {
            kind: ExpressionKind::Literal(literal),
        }
    }
}

#[derive(Debug, PartialEq)]
pub enum ExpressionKind {
    Literal(Literal),
    Binary(Box<Expression>, BinaryOperator, Box<Expression>),
    Unary(UnaryOperator, Box<Expression>),
}
