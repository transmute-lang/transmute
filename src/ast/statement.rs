use crate::ast::expression::ExprId;
use crate::ast::identifier::Identifier;
use crate::lexer::Span;
use std::fmt::{Display, Formatter};

#[derive(Debug, PartialEq)]
pub struct Statement<'s> {
    id: StmtId,
    // todo should be merged with Expression, maybe?
    kind: StatementKind<'s>,
    span: Span,
}

impl<'s> Statement<'s> {
    pub fn new(id: StmtId, kind: StatementKind<'s>, span: Span) -> Self {
        Self { id, kind, span }
    }

    pub fn id(&self) -> StmtId {
        self.id
    }

    pub fn kind(&self) -> &StatementKind<'s> {
        &self.kind
    }
}

#[derive(Debug, Clone, Copy, Eq, PartialEq)]
pub struct StmtId {
    id: usize,
}

impl StmtId {
    pub fn id(&self) -> usize {
        self.id
    }
}

impl Display for StmtId {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.id)
    }
}

impl From<usize> for StmtId {
    fn from(id: usize) -> Self {
        Self { id }
    }
}

#[derive(Debug, PartialEq)]
pub enum StatementKind<'s> {
    Expression(ExprId),
    Let(Identifier<'s>, ExprId),
    Ret(ExprId),
    // todo vec does not hold span and position as it should (search for "Vec<Statement>, Span")
    LetFn(Identifier<'s>, Vec<Identifier<'s>>, Vec<StmtId>),
}
