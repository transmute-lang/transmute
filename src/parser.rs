use crate::ast::expression::{Expression, ExpressionKind};
use crate::ast::identifier::Identifier;
use crate::ast::literal::{Literal, LiteralKind};
use crate::ast::operators::{BinaryOperator, BinaryOperatorKind, UnaryOperator, UnaryOperatorKind};
use crate::ast::statement::{Statement, StatementKind};
use crate::ast::Ast;
use crate::error::Error;
use crate::lexer::{Lexer, PeekableLexer, Span, TokenKind};

pub struct Parser<'a> {
    lexer: PeekableLexer<'a>,
}

impl<'a> Parser<'a> {
    pub fn new(lexer: Lexer<'a>) -> Self {
        Self {
            lexer: PeekableLexer::new(lexer, 1),
        }
    }

    pub fn parse(&mut self) -> Result<Ast, Error> {
        let mut statements = Vec::new();
        while let Ok(token) = self.lexer.peek() {
            if token.kind() == &TokenKind::Eof {
                break;
            }

            statements.push(self.parse_statement()?);
        }

        Ok(Ast::new(statements))
    }

    /// Parses the following:
    /// ```
    /// 'let ident = expr ;
    /// 'let ident ( expr , ... ) = expr ;
    /// 'let ident ( expr , ... ) = { expr ; ... }
    /// 'expr ;
    /// 'ret expr ;
    /// ```
    fn parse_statement(&mut self) -> Result<Statement, Error> {
        // fixme: duplicated several times
        let token = match self.lexer.peek() {
            Ok(token) => token,
            Err(_) => Err(self.lexer.next().unwrap_err())?,
        };

        match token.kind() {
            TokenKind::Let => {
                let let_token = self.lexer.next().expect("we just peeked it");
                let identifier_token = self.lexer.next()?;
                let identifier = match identifier_token.kind() {
                    TokenKind::Identifier(ident) => {
                        Identifier::new(ident.to_string(), identifier_token.span().clone())
                    }
                    _ => todo!(
                        "parse_statement: error handling {:?} at {:?}",
                        identifier_token.kind(),
                        identifier_token.span()
                    ),
                };

                let token = match self.lexer.peek() {
                    Ok(token) => token,
                    Err(_) => Err(self.lexer.next().unwrap_err())?,
                };

                if token.kind() == &TokenKind::OpenParenthesis {
                    // let ident '( expr , ... ) = expr ;
                    return self.parse_function(let_token.span(), identifier);
                }

                // fixme: pattern duplicated several times
                let equal_token = self.lexer.next()?;
                if equal_token.kind() != &TokenKind::Equal {
                    todo!(
                        "parse_statement: error handling {:?}; expected = at {:?}",
                        equal_token.kind(),
                        equal_token.span()
                    )
                }

                let expression = self.parse_expression()?;

                // fixme: duplicated several times
                let semicolon_token = self.lexer.next()?;
                if semicolon_token.kind() != &TokenKind::Semicolon {
                    todo!(
                        "parse_statement: error handling {:?}; expected ; at {:?}",
                        semicolon_token.kind(),
                        semicolon_token.span()
                    )
                }

                let span = let_token.span().extend_to(semicolon_token.span());
                Ok(Statement::new(
                    StatementKind::Let(identifier, expression),
                    span,
                ))
            }
            TokenKind::Ret => {
                let ret_token = self.lexer.next().expect("we just peeked it");
                let expression = self.parse_expression()?;
                let semicolon_token = self.lexer.next()?;
                if semicolon_token.kind() != &TokenKind::Semicolon {
                    todo!(
                        "parse_statement: error handling {:?}; expected ; at {:?}",
                        semicolon_token.kind(),
                        semicolon_token.span()
                    )
                }

                Ok(Statement::new(
                    StatementKind::Ret(expression),
                    ret_token.span().extend_to(semicolon_token.span()),
                ))
            }
            TokenKind::Minus
            | TokenKind::Number(_)
            | TokenKind::OpenParenthesis
            | TokenKind::Identifier(_) => {
                let expression = self.parse_expression()?;
                let semicolon_token = self.lexer.next()?;
                if semicolon_token.kind() != &TokenKind::Semicolon {
                    todo!(
                        "parse_statement: error handling {:?}; expected ; at {:?}",
                        semicolon_token.kind(),
                        semicolon_token.span()
                    )
                }
                let span = expression.span().extend_to(semicolon_token.span());
                Ok(Statement::new(StatementKind::Expression(expression), span))
            }
            _ => todo!(
                "parse_statement: error handling {:?} at {:?}",
                token.kind(),
                token.span()
            ),
        }
    }

    /// Parses the following:
    /// ```
    /// let ident '( expr , ... ) = expr ;
    /// let ident '( expr , ... ) = { expr ; ... }
    /// ```
    fn parse_function(&mut self, span: &Span, identifier: Identifier) -> Result<Statement, Error> {
        // let name '( expr , ... ) = expr ;
        let open_parenthesis_token = self.lexer.next().expect("present, as we peeked it already");
        assert_eq!(open_parenthesis_token.kind(), &TokenKind::OpenParenthesis);

        let mut parameters = Vec::new();
        let mut comma_seen = true;

        loop {
            let token = self.lexer.next()?;
            match token.kind() {
                TokenKind::CloseParenthesis => {
                    break;
                }
                TokenKind::Comma if !comma_seen => {
                    comma_seen = true;
                }
                TokenKind::Identifier(ident) if comma_seen => {
                    parameters.push(Identifier::new(ident.to_string(), token.span().clone()));
                    comma_seen = false;
                }
                _ => todo!(
                    "parse_function: error handling {:?} at {:?}",
                    token.kind(),
                    token.span()
                ),
            }
        }

        // let name ( expr , ... ) '= expr ;
        let token = self.lexer.next()?;
        if token.kind() != &TokenKind::Equal {
            todo!(
                "parse_function: error handling {:?} at {:?}",
                token.kind(),
                token.span()
            )
        }

        let token = match self.lexer.peek() {
            Ok(token) => token,
            Err(_) => Err(self.lexer.next().unwrap_err())?,
        };

        let (statements, end_span) = if token.kind() == &TokenKind::OpenCurlyBracket {
            // let name ( expr , ... ) = '{ expr ; ... }
            let _open_curly_bracket_token = self.lexer.next().expect("we just peeked it");

            let mut statements = Vec::new();

            let span = loop {
                let token = match self.lexer.peek() {
                    Ok(token) => token,
                    Err(_) => Err(self.lexer.next().unwrap_err())?,
                };
                match token.kind() {
                    TokenKind::CloseCurlyBracket => {
                        let token = self.lexer.next().expect("we just peeked it");
                        break token.span().clone();
                    }
                    _ => {
                        statements.push(self.parse_statement()?);
                    }
                }
            };

            (statements, span)
        } else {
            // let name ( expr , ... ) = 'expr ;
            let expression = self.parse_expression()?;

            let semicolon_token = self.lexer.next()?;
            if semicolon_token.kind() != &TokenKind::Semicolon {
                todo!(
                    "parse_statement: error handling {:?}; expected = at {:?}",
                    semicolon_token.kind(),
                    semicolon_token.span()
                )
            }

            let end_span = expression.span().extend_to(semicolon_token.span());

            (
                vec![Statement::new(
                    StatementKind::Expression(expression),
                    end_span.clone(),
                )],
                end_span,
            )
        };

        Ok(Statement::new(
            StatementKind::LetFn(identifier, parameters, statements),
            span.extend_to(&end_span),
        ))
    }

    /// Parses the following:
    /// ```
    /// 'expr
    /// ```
    fn parse_expression(&mut self) -> Result<Expression, Error> {
        self.parse_expression_with_precedence(Precedence::Lowest)
    }

    fn parse_expression_with_precedence(
        &mut self,
        precedence: Precedence,
    ) -> Result<Expression, Error> {
        let token = self.lexer.next()?;

        let mut expression = match token.kind() {
            TokenKind::Identifier(ident) => {
                let identifier = Identifier::new(ident.clone(), token.span().clone());

                let token = match self.lexer.peek() {
                    Ok(token) => token,
                    Err(_) => Err(self.lexer.next().unwrap_err())?,
                };

                match token.kind() {
                    &TokenKind::OpenParenthesis => self.parse_function_call(identifier)?,
                    _ => {
                        let span = identifier.span().clone();
                        Expression::from(Literal::new(LiteralKind::Identifier(identifier), span))
                    }
                }
            }
            TokenKind::Number(n) => {
                Expression::from(Literal::new(LiteralKind::Number(*n), token.span().clone()))
            }
            TokenKind::Minus => {
                let expression = self.parse_expression_with_precedence(Precedence::Prefix)?;
                let span = token.span().extend_to(expression.span());
                Expression::new(
                    ExpressionKind::Unary(
                        UnaryOperator::new(UnaryOperatorKind::Minus, token.span().clone()),
                        Box::new(expression),
                    ),
                    span,
                )
            }
            TokenKind::OpenParenthesis => {
                let open_loc = token.span();

                let expression = self.parse_expression()?;
                let token = self.lexer.next()?;
                if token.kind() == &TokenKind::CloseParenthesis {
                    expression
                } else {
                    todo!(
                        "parse_expression_with_precedence: error handling {:?}; expected ) at {:?} for parenthesis open at {:?}",
                        token.kind(),
                        token.span(),
                        open_loc
                    )
                }
            }
            _ => todo!(
                "parse_expression_with_precedence: error handling {:?} at {:?}",
                token.kind(),
                token.span()
            ),
        };

        while let Some(next_precedence) = self
            .lexer
            .peek()
            .as_ref()
            .map(|t| Into::<Option<Precedence>>::into(t.kind()))
            .unwrap_or_default()
        {
            if precedence < next_precedence {
                expression = self.parse_infix_expression(expression, next_precedence)?;
            } else {
                break;
            }
        }

        Ok(expression)
    }

    /// Parses the following:
    /// ```
    /// identifier ( expr , ... )
    /// ```
    fn parse_function_call(&mut self, identifier: Identifier) -> Result<Expression, Error> {
        // identifier '( expr , ... )
        let open_parenthesis_token = self.lexer.next()?;
        assert_eq!(open_parenthesis_token.kind(), &TokenKind::OpenParenthesis);

        let mut arguments = Vec::new();
        let mut comma_seen = true;

        let span = loop {
            let token = match self.lexer.peek() {
                Ok(token) => token,
                Err(_) => Err(self.lexer.next().unwrap_err())?,
            };
            match *token.kind() {
                TokenKind::CloseParenthesis => {
                    let token = self.lexer.next().expect("we just peeked it");
                    break identifier.span().extend_to(token.span());
                }
                TokenKind::Comma if !comma_seen => {
                    self.lexer.next().expect("we just peeked it");
                    comma_seen = true;
                }
                TokenKind::Comma if comma_seen => {
                    todo!(
                        "parse_function_call: error handling {:?} at {:?}",
                        token.kind(),
                        token.span()
                    )
                }
                _ if comma_seen => {
                    arguments.push(self.parse_expression()?);
                    comma_seen = false;
                }
                _ => todo!(
                    "parse_function_call: error handling {:?} at {:?}",
                    token.kind(),
                    token.span()
                ),
            }
        };

        // identifier ( expr , ... ) '

        Ok(Expression::new(
            ExpressionKind::MethodCall(identifier, arguments),
            span,
        ))
    }

    fn parse_infix_expression(
        &mut self,
        left: Expression,
        precedence: Precedence,
    ) -> Result<Expression, Error> {
        let operator = self
            .parse_operator()
            .expect("there must be an operator, we got a precedence");
        let right = self.parse_expression_with_precedence(precedence)?;

        let span = left.span().extend_to(right.span());

        Ok(Expression::new(
            ExpressionKind::Binary(Box::new(left), operator, Box::new(right)),
            span,
        ))
    }

    fn parse_operator(&mut self) -> Result<BinaryOperator, Error> {
        let token = self.lexer.next()?;

        match token.kind() {
            TokenKind::Plus => Ok(BinaryOperator::new(
                BinaryOperatorKind::Addition,
                token.span().clone(),
            )),
            TokenKind::Minus => Ok(BinaryOperator::new(
                BinaryOperatorKind::Subtraction,
                token.span().clone(),
            )),
            TokenKind::Star => Ok(BinaryOperator::new(
                BinaryOperatorKind::Multiplication,
                token.span().clone(),
            )),
            TokenKind::Slash => Ok(BinaryOperator::new(
                BinaryOperatorKind::Division,
                token.span().clone(),
            )),
            _ => todo!(
                "parse_operator: error handling {:?} at {:?}",
                token.kind(),
                token.span()
            ),
        }
    }
}

#[derive(Debug, Ord, PartialOrd, Eq, PartialEq)]
enum Precedence {
    Lowest,
    Sum,
    Product,
    Prefix,
}

impl From<&TokenKind> for Option<Precedence> {
    fn from(kind: &TokenKind) -> Self {
        match kind {
            TokenKind::Plus => Some(Precedence::Sum),
            TokenKind::Minus => Some(Precedence::Sum),
            TokenKind::Star => Some(Precedence::Product),
            TokenKind::Slash => Some(Precedence::Product),
            _ => None,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::ast::Ast;
    use crate::lexer::Span;

    #[test]
    fn expression_literal_number() {
        let lexer = Lexer::new("42");
        let mut parser = Parser::new(lexer);

        let actual = parser.parse_expression().unwrap();
        let expected = Expression::new(
            ExpressionKind::Literal(Literal::new(LiteralKind::Number(42), Span::new(1, 1, 0, 2))),
            Span::new(1, 1, 0, 2),
        );

        assert_eq!(expected, actual);
    }

    #[test]
    fn expression_literal_identifier() {
        let mut parser = Parser::new(Lexer::new("forty_two"));

        let actual = parser.parse_expression().unwrap();
        let expected = Expression::new(
            ExpressionKind::Literal(Literal::new(
                LiteralKind::Identifier(Identifier::new(
                    "forty_two".to_string(),
                    Span::new(1, 1, 0, 9),
                )),
                Span::new(1, 1, 0, 9),
            )),
            Span::new(1, 1, 0, 9),
        );

        assert_eq!(actual, expected);
    }

    #[test]
    fn expression() {
        let lexer = Lexer::new("2 + 20 * 2");
        let mut parser = Parser::new(lexer);

        let actual = parser.parse_expression().unwrap();
        let expected = Expression::new(
            ExpressionKind::Binary(
                Box::new(Expression::from(Literal::new(
                    LiteralKind::Number(2),
                    Span::new(1, 1, 0, 1),
                ))),
                BinaryOperator::new(BinaryOperatorKind::Addition, Span::new(1, 3, 2, 1)),
                Box::new(Expression::new(
                    ExpressionKind::Binary(
                        Box::new(Expression::from(Literal::new(
                            LiteralKind::Number(20),
                            Span::new(1, 5, 4, 2),
                        ))),
                        BinaryOperator::new(
                            BinaryOperatorKind::Multiplication,
                            Span::new(1, 8, 7, 1),
                        ),
                        Box::new(Expression::from(Literal::new(
                            LiteralKind::Number(2),
                            Span::new(1, 10, 9, 1),
                        ))),
                    ),
                    Span::new(1, 5, 4, 6),
                )),
            ),
            Span::new(1, 1, 0, 10),
        );

        assert_eq!(actual, expected);
    }

    #[test]
    fn binary_operator_minus() {
        let lexer = Lexer::new("43 - 1");
        let mut parser = Parser::new(lexer);

        let actual = parser.parse_expression().unwrap();
        let expected = Expression::new(
            ExpressionKind::Binary(
                Box::new(Expression::from(Literal::new(
                    LiteralKind::Number(43),
                    Span::new(1, 1, 0, 2),
                ))),
                BinaryOperator::new(BinaryOperatorKind::Subtraction, Span::new(1, 4, 3, 1)),
                Box::new(Expression::from(Literal::new(
                    LiteralKind::Number(1),
                    Span::new(1, 6, 5, 1),
                ))),
            ),
            Span::new(1, 1, 0, 6),
        );

        assert_eq!(actual, expected);
    }

    #[test]
    fn prefix_minus() {
        let mut parser = Parser::new(Lexer::new("- 40 * 2"));

        let actual = parser.parse_expression().expect("expression is valid");
        let expected = Expression::new(
            ExpressionKind::Binary(
                Box::new(Expression::new(
                    ExpressionKind::Unary(
                        UnaryOperator::new(UnaryOperatorKind::Minus, Span::new(1, 1, 0, 1)),
                        Box::new(Expression::new(
                            ExpressionKind::Literal(Literal::new(
                                LiteralKind::Number(40),
                                Span::new(1, 3, 2, 2),
                            )),
                            Span::new(1, 3, 2, 2),
                        )),
                    ),
                    Span::new(1, 1, 0, 4),
                )),
                BinaryOperator::new(BinaryOperatorKind::Multiplication, Span::new(1, 6, 5, 1)),
                Box::new(Expression::new(
                    ExpressionKind::Literal(Literal::new(
                        LiteralKind::Number(2),
                        Span::new(1, 8, 7, 1),
                    )),
                    Span::new(1, 8, 7, 1),
                )),
            ),
            Span::new(1, 1, 0, 8),
        );

        assert_eq!(actual, expected);
    }

    #[test]
    fn root() {
        let lexer = Lexer::new("40 + 2;");
        let mut parser = Parser::new(lexer);

        let actual = parser.parse().unwrap();
        let expected = Ast::new(vec![Statement::new(
            StatementKind::Expression(Expression::new(
                ExpressionKind::Binary(
                    Box::new(Expression::from(Literal::new(
                        LiteralKind::Number(40),
                        Span::new(1, 1, 0, 2),
                    ))),
                    BinaryOperator::new(BinaryOperatorKind::Addition, Span::new(1, 4, 3, 1)),
                    Box::new(Expression::from(Literal::new(
                        LiteralKind::Number(2),
                        Span::new(1, 6, 5, 1),
                    ))),
                ),
                Span::new(1, 1, 0, 6),
            )),
            Span::new(1, 1, 0, 7),
        )]);

        assert_eq!(actual, expected);
    }

    #[test]
    #[should_panic]
    fn missing_right_parenthesis() {
        let mut parser = Parser::new(Lexer::new("(42"));
        let _ = parser.parse();
    }

    #[test]
    fn let_statement() {
        let mut parser = Parser::new(Lexer::new("let forty_two = 42;"));

        let actual = parser.parse_statement().expect("statement is valid");
        let expected = Statement::new(
            StatementKind::Let(
                Identifier::new("forty_two".to_string(), Span::new(1, 5, 4, 9)),
                Expression::new(
                    ExpressionKind::Literal(Literal::new(
                        LiteralKind::Number(42),
                        Span::new(1, 17, 16, 2),
                    )),
                    Span::new(1, 17, 16, 2),
                ),
            ),
            Span::new(1, 1, 0, 19),
        );

        assert_eq!(actual, expected);
    }

    #[test]
    fn two_statements() {
        let mut parser = Parser::new(Lexer::new("let forty_two = 42; forty_two;"));

        let actual = parser.parse().expect("source is valid");
        let expected = Ast::new(vec![
            Statement::new(
                StatementKind::Let(
                    Identifier::new("forty_two".to_string(), Span::new(1, 5, 4, 9)),
                    Expression::new(
                        ExpressionKind::Literal(Literal::new(
                            LiteralKind::Number(42),
                            Span::new(1, 17, 16, 2),
                        )),
                        Span::new(1, 17, 16, 2),
                    ),
                ),
                Span::new(1, 1, 0, 19),
            ),
            Statement::new(
                StatementKind::Expression(Expression::new(
                    ExpressionKind::Literal(Literal::new(
                        LiteralKind::Identifier(Identifier::new(
                            "forty_two".to_string(),
                            Span::new(1, 21, 20, 9),
                        )),
                        Span::new(1, 21, 20, 9),
                    )),
                    Span::new(1, 21, 20, 9),
                )),
                Span::new(1, 21, 20, 10),
            ),
        ]);

        assert_eq!(actual, expected);
    }

    #[test]
    fn function_statement() {
        let mut parser = Parser::new(Lexer::new("let times_two(a) = a * 2;"));

        let actual = parser.parse().expect("source is valid");
        let expected = Ast::new(vec![Statement::new(
            StatementKind::LetFn(
                Identifier::new("times_two".to_string(), Span::new(1, 5, 4, 9)),
                vec![Identifier::new("a".to_string(), Span::new(1, 15, 14, 1))],
                vec![Statement::new(
                    StatementKind::Expression(Expression::new(
                        ExpressionKind::Binary(
                            Box::new(Expression::new(
                                ExpressionKind::Literal(Literal::new(
                                    LiteralKind::Identifier(Identifier::new(
                                        "a".to_string(),
                                        Span::new(1, 20, 19, 1),
                                    )),
                                    Span::new(1, 20, 19, 1),
                                )),
                                Span::new(1, 20, 19, 1),
                            )),
                            BinaryOperator::new(
                                BinaryOperatorKind::Multiplication,
                                Span::new(1, 22, 21, 1),
                            ),
                            Box::new(Expression::new(
                                ExpressionKind::Literal(Literal::new(
                                    LiteralKind::Number(2),
                                    Span::new(1, 24, 23, 1),
                                )),
                                Span::new(1, 24, 23, 1),
                            )),
                        ),
                        Span::new(1, 20, 19, 5),
                    )),
                    Span::new(1, 20, 19, 6),
                )],
            ),
            Span::new(1, 1, 0, 25),
        )]);

        assert_eq!(actual, expected);
    }

    #[test]
    fn function_statements() {
        let mut parser = Parser::new(Lexer::new("let times_two(a) = { a * 2; }"));

        let actual = parser.parse().expect("source is valid");
        let expected = Ast::new(vec![Statement::new(
            StatementKind::LetFn(
                Identifier::new("times_two".to_string(), Span::new(1, 5, 4, 9)),
                vec![Identifier::new("a".to_string(), Span::new(1, 15, 14, 1))],
                vec![Statement::new(
                    StatementKind::Expression(Expression::new(
                        ExpressionKind::Binary(
                            Box::new(Expression::new(
                                ExpressionKind::Literal(Literal::new(
                                    LiteralKind::Identifier(Identifier::new(
                                        "a".to_string(),
                                        Span::new(1, 22, 21, 1),
                                    )),
                                    Span::new(1, 22, 21, 1),
                                )),
                                Span::new(1, 22, 21, 1),
                            )),
                            BinaryOperator::new(
                                BinaryOperatorKind::Multiplication,
                                Span::new(1, 24, 23, 1),
                            ),
                            Box::new(Expression::new(
                                ExpressionKind::Literal(Literal::new(
                                    LiteralKind::Number(2),
                                    Span::new(1, 26, 25, 1),
                                )),
                                Span::new(1, 26, 25, 1),
                            )),
                        ),
                        Span::new(1, 22, 21, 5),
                    )),
                    Span::new(1, 22, 21, 6),
                )],
            ),
            Span::new(1, 1, 0, 29),
        )]);

        assert_eq!(actual, expected);
    }

    #[test]
    fn function_call() {
        let mut parser = Parser::new(Lexer::new("times_two(21)"));
        let actual = parser.parse_expression().expect("expression is valid");
        let expected = Expression::new(
            ExpressionKind::MethodCall(
                Identifier::new("times_two".to_string(), Span::new(1, 1, 0, 9)),
                vec![Expression::new(
                    ExpressionKind::Literal(Literal::new(
                        LiteralKind::Number(21),
                        Span::new(1, 11, 10, 2),
                    )),
                    Span::new(1, 11, 10, 2),
                )],
            ),
            Span::new(1, 1, 0, 13),
        );
        assert_eq!(actual, expected);
    }

    #[test]
    fn ret() {
        let mut parser = Parser::new(Lexer::new("ret 42;"));
        let actual = parser.parse_statement().expect("statement is valid");
        let expected = Statement::new(
            StatementKind::Ret(Expression::new(
                ExpressionKind::Literal(Literal::new(
                    LiteralKind::Number(42),
                    Span::new(1, 5, 4, 2),
                )),
                Span::new(1, 5, 4, 2),
            )),
            Span::new(1, 1, 0, 7),
        );
        assert_eq!(actual, expected);
    }
}
