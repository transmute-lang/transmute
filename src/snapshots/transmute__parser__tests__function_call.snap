---
source: src/parser.rs
expression: "Parser::new(Lexer::new(\"times_two(21)\")).parse()"
---
Err(
    Diagnostics {
        diagnostics: [
            Diagnostic {
                message: "Unexpected `eof`, expected one of `;`, `*`, `/`, `-`, `+`, `==`, `!=`, `>`, `>=`, `<`, `<=`",
                span: 1:14; [13, 13],
                level: Error,
                generated_at: (
                    "src/parser.rs",
                    1163,
                ),
            },
        ],
    },
)
